react-intl/index.d.ts

```diff
@@ -14,129 +12,102 @@ declare namespace ReactIntl {
+    type LocaleData = Locale[];
 
-    function addLocaleData(data: Locale[] | Locale): void;
+    interface InjectIntlConfig {
+        intlPropName?: string;
+        withRef?: boolean;
+    }
 
-    function hasLocaleData(localeName: string): boolean;
+    type ComponentConstructor<P> = React.ComponentClass<P> | React.StatelessComponent<P>
+
+    function injectIntl<P>(component: ComponentConstructor<P & InjectedIntlProps>, options?: InjectIntlConfig): React.ComponentClass<P & InjectedIntlProps>;
+
+    function addLocaleData(data: Locale[] | Locale): void;
 
     interface Messages {
         [key: string]: FormattedMessage.MessageDescriptor
     }
 
     function defineMessages<T extends Messages>(messages: T): T;
 
-    interface IntlShape extends React.Requireable<any> {
+    interface IntlConfig {
+        locale: React.Requireable<any>;
+        formats: React.Requireable<any>;
+        messages: React.Requireable<any>;
+        defaultLocale: React.Requireable<any>;
+        defaultFormats: React.Requireable<any>;
     }
 
-    var intlShape: IntlShape;
+    interface IntlFormat {
+        formatDate: React.Requireable<any>;
+        formatTime: React.Requireable<any>;
+        formatRelative: React.Requireable<any>;
+        formatNumber: React.Requireable<any>;
+        formatPlural: React.Requireable<any>;
+        formatMessage: React.Requireable<any>;
+        formatHTMLMessage: React.Requireable<any>;
+    }
 
-    interface FormatConfig {
-        locale?: string;
-        formats?: any;
+    interface IntlShape extends IntlConfig, IntlFormat, React.Requireable<any> {
+        now: React.Requireable<any>;
     }
 
+    const intlShape: IntlShape;
+
     interface InjectedIntl {
         formatDate: (date: Date, options?: FormattedDate.PropsBase) => string;
         formatTime: (date: Date, options?: FormattedTime.PropsBase) => string;
-        formatRelative: (value: number, options?: FormattedRelative.PropsBase) => string;
+        formatRelative: (value: number, options?: FormattedRelative.PropsBase & { now?: any }) => string;
         formatNumber: (value: number, options?: FormattedNumber.PropsBase) => string;
-        formatPlural: (value: number, options?: FormattedPlural.PropsBase) => string;
-        formatMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: Object) => string;
-        formatHTMLMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: Object) => string;
+        formatPlural: (value: number, options?: FormattedPlural.Base) => keyof FormattedPlural.PropsBase;
+        formatMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: any) => string;
+        formatHTMLMessage: (messageDescriptor: FormattedMessage.MessageDescriptor, values?: any) => string;
     }
 
     interface InjectedIntlProps {
-        intl: InjectedIntl
+        intl?: InjectedIntl
```

Comment:
If you want to use this interface for props then I assume you decorate the component with `injectIntl` first before using. If you don't and still use it the property `intl`will be undefined anyway.

The case for making this optional is to use it in some component such as:
```typescript
interface CustomProps {
  extra: string
}

type FooProps = CustomProps & InjectedIntlProps

const FooComponent: StatelessComponent<FooProps> = ({ intl : { formatMessage } }) => {
  return (<span>{ formatMessage({ id : "foo.bar" }) }</span>)
}

const Foo = injectIntl(FooComponent)
```
Then if you just try to render it within some other render:
```typescript
renderToString(<Foo extra="fooBar" />)
```
This will throw an error as the `intl` is required and not present at the time of render. Component expects to receive the injected props do be defined like `<Foo extra="fooBar" intl={ ... } />`. 

At this point you only care about your custom props and let the decorator take care of `intl` props. At least that was I think. 

This pattern is commonly used for many libraries that inject properties using HoC.